github-api.php

```diff
@@ -3201,71 +3201,388 @@ function vipgoci_github_prs_commits_list(
+			'%s -C %s diff %s %s 2>&1',
+			escapeshellcmd( 'git' ),
+			escapeshellarg( $local_git_repo ),
+			escapeshellarg( $commit_id_a ),
+			escapeshellarg( $commit_id_b )
+		);
 
-		// FIXME: Error-handling
-		$resp_raw = json_decode(
-			vipgoci_github_fetch_url(
-				$github_url,
-				$github_token
+		vipgoci_log(
+			'Running git...',
+			array(
+				'cmd'	=> $git_diff_cmd
 			)
 		);
 
+		/* Actually execute */
+		vipgoci_runtime_measure( VIPGOCI_RUNTIME_START, 'git_cli' );
+
+		$git_diff_results = shell_exec( $git_diff_cmd );
+
+		vipgoci_runtime_measure( VIPGOCI_RUNTIME_STOP, 'git_cli' );
+
+		/*
+		 * Prepare intermediate results array.
+		 */
+		$diff_results = array(
+			'files' => array()
+		);
+
+		/*
+		 * Split results into array
+		 */
+
+		$git_diff_results = explode(
+			PHP_EOL,
+			$git_diff_results
+		);
+
+		/*
+		 * Initialize stateful variables.
+		 */
+
+		$cur_file = null;
+		$cur_mode = 'info'; // Other mode is 'patch'
+		$cur_file_first_patch_line = true;
+		$cur_file_status = null;
+		$cur_file_minus = null;
+		$cur_file_plus = null;
+		$cur_file_paths_cleaned = false;
+		$cur_file_plus_plus_plus_seen = false;
+		$cur_file_patch_buffer = '';
+
+		/*
+		 * Process the output of git diff, one line
+		 * at a time. Update state-variables as we go
+		 * so we can keep track of what is happening.
+		 *
+		 * The output looks like this:
+		 * 
+		 *  diff --git a/renamed-file2.txt b/renamed-file2.txt
+		 *  deleted file mode 100644
+		 *  index 55ab87b..0000000
+		 *  --- a/renamed-file2.txt
+		 *  +++ /dev/null
+		 *  @@ -1,2 +0,0 @@
+		 *  -# vip-go-ci-testing
+		 *  -Pull-Requests, commits a
+		 *  diff --git ...
+		 *  ...
+		 */
+
+		foreach( $git_diff_results as $git_result_item ) {
+			/*
+			 * Split each line into array at spaces,
+			 * making it easy to process each line of results.
+			 */
+			$git_result_item_arr = explode(
+				' ',
+				$git_result_item
+			);
+
+			/*
+			 * Check if we are seeing 'diff --git ... 
+			 * and if so switch to info mode, get file names
+			 * and set other state variables.
+			 */
+			if (
+				( ! empty( $git_result_item_arr ) ) &&
+				( 'diff' === $git_result_item_arr[0] ) &&
+				( '--git' === $git_result_item_arr[1] )
+			) {
+				$cur_mode = 'info';
+	
+				$cur_file = null;
+				$cur_file_status = null;
+
+				$cur_file_minus = $git_result_item_arr[2];
+				$cur_file_plus = $git_result_item_arr[3];
+
+				$cur_file_paths_cleaned = false;
+				$cur_file_plus_plus_plus_seen = false;
+
+				$cur_file_patch_buffer = '';
+			}
+
+			else if (
+				( 'info' === $cur_mode ) &&
+				( ! empty( $git_result_item_arr ) )
+			) {
+				if (
+					( 'new' === $git_result_item_arr[0] ) &&
+					( 'mode' === $git_result_item_arr[1] )
+				) {
+					$cur_file_status = 'modified';
+				}
+
+				else if (
+					( 'rename' === $git_result_item_arr[0] ) &&
+					( 'to' === $git_result_item_arr[1] )
+				) {
+					$cur_file_plus_plus_plus_seen = true;
+				}
+
+				else if (
```

Comment:
We can also refactor a little bit this [block](https://github.com/Automattic/vip-go-ci/pull/153/files#diff-4a0fcb15c05fb7f32eb99b19cb448971027c14ca8f088625a92a9e537795c0ffR3363-R3376)

```
                else if (
                    ( '---' === $git_result_item_arr[0] )
                ) {
                    $cur_file_minus = $git_result_item_arr[1];
                    $cur_file_paths_cleaned = false;
                }
                
                else if ( '+++' === $git_result_item_arr[0] ) {
                    $cur_file_plus = $git_result_item_arr[1];
                    $cur_file_plus_plus_plus_seen = true;
                    $cur_file_paths_cleaned = false;
                }
```

Reply:
I agree. I will change this.

Reply:
Fixed in bd9770c4d1bbf9a4a4d2f31f2c89774313226def